### ISSUE TITLE
Incorrect username pattern validation with i18n support enabled

### ISSUE THREAD

**Issue by hpadmanabhan** [jhipster/generator-jhipster#5432]
##### **Overview of the issue**

Pattern based Input validation for username on the registration page on projects generated with i18n support enabled is incorrect.

It limits input to English alphabet letters only instead of allowing letters in the alphabet of the selected language.

##### **Motivation for or Use Case**

If I enable support for French/Spanish/German/Russian/etc and a user selects one of those languages and decides to register, that user should be able to enter a username using the alphabet of the selected language.

##### **Reproduce the error**

- Generate a project with support enabled for say Russian
- Switch language to Russian on the UI
- Attempt to register with username 'Русский'
- Receive error message saying "Ваш логин должен содержать только буквы в нижнем регистре и цифры"
- On a  side note, the error message does not match the pattern matched, the pattern allows for some special characters and upper case letters as well, however, the error message states otherwise.
- Another side note, I believe the password strength bar also suffers from this issue and will incorrectly indicate the password as being weak if non-English characters are used.

##### **Related issues**

None that I am aware of.

##### **Suggest a Fix**

The issue is with the following static pattern usage for validation:
`ng-pattern="/^[_'.@A-Za-z0-9-]*$/"`

Javascript unfortunately is in the stone ages as far as unicode validations go so using ng-pattern for validation across languages is not an option.
Fortunately [xregexp](https://github.com/slevithan/xregexp) provides a way out. I have added the following custom directive wrapping xregexp as the solution in my project since I need unicode based pattern validations in multiple places:
`
(function() {
    'use strict';

    angular
        .module('myApp')
        .directive('xregexpPattern', xregexpPattern);
    
    function xregexpPattern () {
    	var directive = {
    			restrict: 'A',
    			require: 'ngModel',
    			link: linkFunc
    	};
    	return directive;
    	
    	function linkFunc (scope, element, attr, ctrl) {
    		var regex = new XRegExp(attr.xregexpPattern);

    		function validateModel(ngModelValue) {
        		if (regex.test(ngModelValue)) {
        			ctrl.$setValidity('xregexpPattern', true);
        		}
        		else {
        			ctrl.$setValidity('xregexpPattern', false);
        		}
        		return ngModelValue;
    		}
    		
    		ctrl.$parsers.push(validateModel);
    	}
    }
})();

`

I switch out the ng-pattern directive in the input field with the custom directive:
`xregexp-pattern="{{vm.namePattern}}"`
With the pattern being specified in the controller as:
`vm.namePattern = '^[_'.@\\p{L}\\p{Nd}-]*$';`

This solution requires adding xregexp dependency to the project:
`bower install xregexp --save`

For now the above uses a static pattern allowing all unicode letters, I need to link it up with $translate.use() to restrict it to the currently selected language with a filter mapping languages to the corresponding character set constants.

This solution or something along these lines needs to be put in place or at a minimum the existing pattern needs to be disabled when a project is generated with i18n support since it is incorrect.

##### **JHipster Version(s)**

4.x

**Comment by deepu105** [comment]
I dont want to add a library just for this. I would rather just do an ascii char validation like mentioned here http://stackoverflow.com/questions/3203190/regex-any-ascii-character

**Comment by hpadmanabhan** [comment]
I'm going to assume its been a while from CS101 for you and you have forgotten what language alphabets are included in ASCII and that you didn't bother to test this 'solution' with the reproduction example cited above.
ASCII will barely add a few additional characters like umlaut and some unnecessary special characters but is otherwise not too different from what you already have, which is essentially the basic latin alphabets.

Shouldn't the question be what the end users want?

As it stands your implementation is incorrect and limiting in its validation from advertised support and gives users a false impression of what can be done. 
If you don't want to actually add support for advertised languages then at least get rid of automatically adding such limiting validations that will only frustrate end users (and mislead them in the case of password strength). 

I love what you guys have done with JHipster so far and I would think this would fit exactly into what you want from it. As it  stands now it would require people that use JHipster and want actual language support to have to rewrite those incorrect validations with boilerplate code like I have, precisely what JHipster is supposed to take care of, unless I am mistaken.

**Comment by deepu105** [comment]
@hpadmanabhan I do know what ASCII and Unicode is, it was just a hint towards something that be done without using a library. And ofcourse I dont test every comment I just make as its just an idea and when implementing it the implementer needs to work on the idea. btw it is possible with similar approach check this http://stackoverflow.com/questions/150033/regular-expression-to-match-non-english-characters/873600#873600

Now for your next question yes and no. Plz remember that this is an OSS free project which we all work on during our free time and hence implementing anything will need to satisfy things like value provided, maintainability, code complexity etc and not just what users what. So I reiterate that we do not want to integrate another library and write a directive just for this.

Now back to the issue. Its not very common to have usernames in non latin characters and thats why this is the first time we get a bug report for this.  So no this is not a bug it was a decision to support only latin alphabets without any special characters. If your business requirement requires more ofcourse you need to make the code changes. Now a solution for you could be to use a regex that only negates the characters you want to exclude in user names. But you also need to take things like compatibility with spring security etc for use of these languages. I think there was a specific reason for the decision but I couldnt recall it @jhipster/developers any body remember why we restricted username like this?

**Comment by jdubois** [comment]
@hpadmanabhan I'm sorry but I don't allow any insults on team members here. We're doing this on our free time, and I *hate* to have to handle stuff like, during the week-end.
Nobody's going to help you for free if you are not nice with them.

**Comment by hpadmanabhan** [comment]
I apologize if it came across as an insult as it wasn't meant to be, honestly his initial response came across as dismissive and not concerned with the facts of the problem at all unlike the second response.
I appreciate what you are doing which is why I came back with the issue and potential solutions for it instead of keeping it to myself.
I already have a resolution for the problem, my only goal here was to help JHipster and other users of JHipster, feel free to be combative and dismissive of it.

**Comment by hpadmanabhan** [comment]
@deepu105 here is a SO thread that answers most of your questions and probably better explains a bunch of mine:
https://stackoverflow.com/questions/150033/regular-expression-to-match-non-english-characters

A unicode regular expressions primer if it helps:
http://www.regular-expressions.info/unicode.html

**Comment by hpadmanabhan** [comment]
The goal is to have the same restrictions as the existing validation (limit to letter characters only), simply extended to other supported languages and there is no clean way to do so in JS except redoing what XRegExp does and the nice thing about this library is that it supports the POSIX sytax like most other languages which JS doesn't.

**Comment by jdubois** [comment]
@deepu105 just to give some history on this decision: I originally only coded this as an example, and thought people would indeed make their own regexp, depending on their business requirements. This is just a simple example and/or a placeholder, nothing more.

**Comment by hpadmanabhan** [comment]
@jdubois That's what I figured and I think its fine as such, probably good to indicate it in the docs as well so folks don't assume otherwise. As @deepu105 points out, it probably works fine in most use cases which are simple and English focused.
My idea here was to extend that and provide an option for those interested in Unicode support to have a way of simply specifying expressions in POSIX syntax instead of having to write all of the biolerplate required to support it.

**Comment by hpadmanabhan** [comment]
@deepu105 Java unlike JS is unicode by default, there are no issues with Spring Security with this, however, databases are another matter, you need to enable  UTF-8 charset in most explicitly when creating databases, but then, that should be required up front when enabling i18n, again, something that should perhaps be indicated by the generator or through Liquibase settings.

**Comment by deepu105** [comment]
@jdubois dont worry it will take a lot more to make me feel offended :smiling_imp: 
@hpadmanabhan I dont need a class on regexes or character encoding, I'm pretty familiar with both. And incase you didnt bother to read I posted the same links you did above right before that (I can google as well you see)
Let me say it again we can't and won't handle every business use case out there our goal  is to provide something for the users to start with and not to do entire coding for you and I think thats made pretty clear in our docs. We have not mentioned that we allow user id creation in any language, and the error message throws is clear I guess (atleast in english) well we could mention that we dont allow but then we would be here writing docs on everything that we dont do. To be fair if you generate an app I would expect you to go through the code and understand it and make the required adjustments for your use case.

So in short nope we wont be handling this by using another library and for us this is low priority. Having said that if you want to improve this you are welcome to do a PR as long as its a simple regex solution. you could also consider adding this as a [TIP](https://jhipster.github.io/tips/) on our docs website or even building a plugin [module](https://jhipster.github.io/modules/creating-a-module/) for this

**Comment by SorinaAvram** [comment]
Hemant Padmanabhan !!
